internal/pkg/provision/providers/qemu/launch.go

```diff
@@ -0,0 +1,119 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package qemu
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/talos-systems/talos/internal/pkg/provision/providers/vm"
+)
+
+// LaunchConfig is passed in to the Launch function over stdin.
+type LaunchConfig struct {
+	DiskPath        string
+	VCPUCount       int64
+	MemSize         int64
+	QemuExecutable  string
+	KernelImagePath string
+	InitrdPath      string
+	KernelArgs      string
+}
+
+// Launch a control process around qemu VM manager.
+//
+// This function is invoked from 'talosctl qemu-launch' hidden command
+// and wraps starting, controlling 'qemu' VM process.
+//
+// Launch restarts VM forever until control process is stopped itself with a signal.
+//
+// Process is expected to receive configuration on stdin. Current working directory
+// should be cluster state directory, process output should be redirected to the
+// logfile in state directory.
+//
+// When signals SIGINT, SIGTERM are received, control process stops qemu and exits.
+//
+//nolint: gocyclo
+func Launch() error {
+	var config LaunchConfig
+
+	if err := vm.ReadConfig(&config); err != nil {
+		return err
+	}
+
+	c := vm.ConfigureSignals()
+
+	args := []string{
+		"-m", strconv.FormatInt(config.MemSize, 10),
+		"-drive", fmt.Sprintf("format=raw,if=virtio,file=%s", config.DiskPath),
+		"-smp", fmt.Sprintf("cpus=%d", config.VCPUCount),
+		"-accel",
+		"kvm",
+		"-nographic",
+	}
+
+	disk, err := os.Open(config.DiskPath)
+	if err != nil {
+		return fmt.Errorf("failed to open disk file %w", err)
+	}
+
+	// check if disk is empty
+	checkSize := 512
+	buf := make([]byte, checkSize)
+
+	_, err = disk.Read(buf)
+	if err != nil {
+		return fmt.Errorf("failed to read disk file %w", err)
+	}
+
+	if bytes.Equal(buf, make([]byte, checkSize)) {
+		args = append(args,
+			"-kernel", config.KernelImagePath,
+			"-initrd", config.InitrdPath,
+			"-append", config.KernelArgs,
+		)
```

Comment:
we might want to add `-no-reboot` here so that `qemu` runs with our kernel until the reboot (as installer should have run and installed the Talos, so next reboot it should go with the bootloader)

Reply:
if we add `-no-reboot`, we might also need to make a big `for` loop here like with firecracker so that qemu is restarted if it exits cleanly

Reply:
gotcha, will make a change in a moment